### PR TITLE
Implement 5x5 puzzle

### PR DESCRIPTION
**Pull request by FrankFeng27**
#### Abstract

This submission made the following changes:
1, on_pushButton_14_clicked() was too complicated, ("Auto" button). We divided its logic to three states: `kInitial, kRunning, kPaused`, and switchAutoState() would process state transition.
2. add spdlog and implement logging to file.
3. add 5 x 5 puzzle.
4. when auto-playing, QTimer::singleShot was used to replace while loop.

Auto-play states diagram:
Statemachine Diagram0

### CODE REVIEW

File: TilePuzzle.pro [hunk 1]
```diff
@@ -8,15 +8,19 @@ CONFIG += c++11
 # In order to do so, uncomment the following line.
 #DEFINES += QT_DISABLE_DEPRECATED_BEFORE=0x060000    # disables all the APIs deprecated before Qt 6.0.0
 
+INCLUDEPATH += includes/
```
Comment (FrankFeng27) [inline TilePuzzle.pro:11]: Support to reference to spdlog library.

File: utilities.cpp [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+#ifndef SPDLOG
+static streambuf* s_recordedBuf = nullptr;
+static ofstream s_out;
+#endif
+
+static bool s_initialized = false;
+
+void utilitiesInitialize(char const * fp) {
+    if (fp == nullptr)
```
Comment (FrankFeng27) [inline utilities.cpp:23]: If log file not be specified, we won't write log to file.

File: utilities.h [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+void utilitiesInitialize(char const * fp=nullptr);
+void utilitiesUninitializePrint();
+
+void printPuzzle(int * puzzle, int dim);
+
+#if defined(SPDLOG)
+#define LOGINFO spdlog::get(LOGNAME)->info
+#else
+#define LOGINFO(x, ...)
```
Comment (FrankFeng27) [inline utilities.h:24]: If SPDLOG was not turn on, LOGINFO would have no effect.

File: mainwindow.cpp [hunk 4]
```diff
@@ -273,105 +371,46 @@ void MainWindow::on_pushButton_13_clicked()
     ui->label_5->setStyleSheet("color:rgb(255,180,241)");
 }
 
+void MainWindow::automate_callback() {
```
Comment (FrankFeng27) [inline mainwindow.cpp:374]: This function replaced the original while loop and QApplication::processEvent() in on_pushButton_14_clicked()

File: mainwindow.cpp [hunk 5]
```diff
@@ -273,105 +371,46 @@ void MainWindow::on_pushButton_13_clicked()
...
+    ui->pushButton_14->setText(stepCountNum);
+    LOGINFO("automate_callback begin ... ");
+    printPuzzle(puzzle, dimension);
+    LOGINFO("stepCount: {}", stepCount);
+    LOGINFO("automate_callback end ... ");
+    showStep(puzzle, steps[stepCount], stepCount, dimension);
+    setText(puzzle, ui, dimension * dimension);
+    setColor(puzzle, ui, dimension * dimension);
+    ui->centralwidget->repaint();
+    QTimer::singleShot(autoSpeed, this, &MainWindow::automate_callback);
```
Comment (FrankFeng27) [inline mainwindow.cpp:395]: emit QTimer event again.

**Comment by FrankFeng27** [comment]
@DancingNeko , when merging to master, better to use `Squash and merge` instead of `Merge pull request`.

**Review by DeFisch — APPROVED** [review]

**Pull request merged**